docs/userguide/joins.rst

```diff
@@ -0,0 +1,260 @@
+
+.. code-block:: python
+
+  post_process_futures = []
+  for patch in patch_list:
+    sensor_list_future = generate_sensor_list_app(patch)
+    sensor_futures = []
+    for sensor in sensor_list_future.result():
+      sensor_futures.append(process_sensor(sensor))
+    assembled_future = assemble_sensors(*sensor_futures)
+    post_process_futures.append(post_process(assembled_future))
+  mosaic_future = make_mosaic(*post_process_futures)
+  mosaic_future.result()
+
+Weaknesses
+""""""""""
+
+*  Only one patch is processed at a time. The outer for loop blocks on task completion repeatedly
+   before all of the apps are submitted, forcing unnecessarily serialised execution ordering: each
+   loop will block waiting for generate_sensor_list_app to complete.
+
+Attempt 2 - a `python_app` per iteration
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Put the whole loop body into an app.
+
+.. code-block:: python
+
+  patch_futures = []
+  for patch in patch_list:
+    patch_futures = process_patch(patch)
+  mosaic_future = make_mosaic(*patch_futures)
+  mosaic_future.result()
+
+  @python_app(executors=['local-thread-pool']
+  def process_patch():
+    sensor_list_future = generate_sensor_list_app(patch)
+    sensor_futures = []
+    for sensor in sensor_list_future.result():
+      sensor_futures.append(process_sensor(sensor))
+    assembled_future = assemble_sensors(*sensor_futures)
+    post_process_future = post_process(assembled_future)
+    post_process_future.result()
+
+In this attempt, all of the for-loop bodies are launched without blocking (as process_patch invocations),
+and potentially can run concurrently.
+
+Weaknesses
+""""""""""
+
+Each process_patch app must occupying a thread pool worker for the entire duration of
+the tasks that it has launched, because it blocks waiting for completion of the post_process step, by
+calling post_process_future.result().
+
+It does this so that the process_patch app completes after the post_process step. If this call to
+.result() was not there, the process_patch app invocation would complete too early, and make_mosaic
+could run before post_process is complete.
+
+Because of this, there can be serialization and deadlock issues: there needs to be one local-thread-pool
+worker available for every patch to be processed simultanously, that will for the most part be sitting
+idle waiting for final results. If there are fewer workers, then patch processing will be
+serialised due to lack of workers, although to a lesser extent than in attempt 1.
+
+But worse, if any of the launched apps also used the same `ThreadPoolExecutor`, then
+the workflow can deadlock:
+launched process_patch apps will be waiting for other apps to complete, but those apps cannot start
+because process_patch apps are occupying all of the `ThreadPoolExecutor` workers.
+
+This leads to a principle: apps should not block on other apps; instead any blocking of execution
+should happen inside parsl's dependency mechanism.
```

Comment:
Is this a more general principle than just in these examples? Something that we should use to rethink some of other tutorials examples?

Reply:
yes. (it was in the parslfest talk on structuring workflows too, i think)

Reply:
and there's even a userguide chapter on structuring workflows that this could form part of.